### PR TITLE
Samsung part customer attestation

### PR DESCRIPTION
**Pull request by masonmcelvain**
Closes https://github.com/iFixit/checkout-ui-extensions/issues/27
Original spec is here https://github.com/iFixit/ifixit/issues/43322

This should only show up on the US store when the app is eventually deployed to production. Therefore nothing is localized and all text is in English.

I discovered that we can't use `useAppMetafields` to get a metafield for all products in the cart due to a limitation in the hooks Shopify provides. More details in https://github.com/Shopify/ui-extensions/issues/1780. In this PR we use the Storefront API to work around the limitation.

## QA

Make sure the OEM partnership attestation renders on the first page of checkout only when samsung genuine parts are in the cart (as determined by the "OEM Partnership JSON" metafield). [Here](https://ifixit-checkout-dev.myshopify.com/products/samsung-galaxy-s22-ultra-usa-screen-and-battery-genuine) is an example product.

Customers should not be able to complete checkout without checking this box.

In shopify-v2, this component is a popup. It was easier to implement this as a checkbox in this framework, not sure if popups are even possible.

| Screenshot |
| --- |
| image |

### CODE REVIEW

**Comment by deltuh-vee** [comment]
Is the `do not impact` link supposed to go [here](https://www.ifixit.com/News/14378/passionate-about-right-to-repair-the-ftc-wants-to-hear-from-you)
CleanShot 2024-02-27 at 10 51 30

**Comment by deltuh-vee** [comment]
dev_block ✊ 
I'm unable to continue through checkout even without a Samsung product present.
CleanShot 2024-02-27 at 10 54 24

**Comment by masonmcelvain** [comment]
It should go to the same place outlined in the [spec](https://github.com/iFixit/ifixit/issues/43322).
https://docs.google.com/document/d/1nLOfzSzHYsrBkiLtzYgZucMKRWHFP7_Y2O24CsKDXWU/edit

**Comment by masonmcelvain** [comment]
> I'm unable to continue through checkout even without a Samsung product present.

This should be addressed now, thanks! un_dev_block 👍🏻

**Comment by masonmcelvain** [comment]
> Is the `do not impact` link supposed to go [here](https://www.ifixit.com/News/14378/passionate-about-right-to-repair-the-ftc-wants-to-hear-from-you)?

I see what you mean now, I asked Scott [here in slack.](https://ifixit.slack.com/archives/C0252FETV2S/p1709073435032339)

deploy_block 👍🏻 on that

**Comment by masonmcelvain** [comment]
un_deploy_block 👍🏻

**Comment by deltuh-vee** [comment]
QA 🎬 
* If a Samsung Genuine Part is in the cart, customers have to check the attestation box to continue with their order.
* Orders without Samsung Genuine Parts continue as normal.

File: extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+   'Please review and agree to the attestation to continue.';
+
+export function SamsungConsent() {
+   const appMetafields = useAppMetafields();
+   const shop = useShop();
+   const [partnershipCodes, setPartnershipCodes] = useState<string[]>([]);
+   const [isChecked, setIsChecked] = useState<boolean>(false);
+   const [showError, setShowError] = useState<boolean>(false);
+   useEffect(() => {
+      setPartnershipCodes(
```
Comment (danielcliu-ifixit) [inline extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx:28]: this is nice, but its not really necessary right now. We could have a state for `hasSamsungProduct` and just do a find in there without saving out to another state.

File: extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+              errors: [{ message: requiredErrorMessage }],
+              perform,
+           }
+         : {
+              behavior: 'allow',
+              perform,
+           }
+   );
+
+   if (!isShowing) return null;
```
Comment (danielcliu-ifixit) [inline extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx:67]: could we just have this return on line 50? or is there a react hooks thing blocking us from doing that?

File: extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      canBlockProgress && isShowing && !isChecked
+         ? {
+              behavior: 'block',
+              reason: 'Attestation required for a product in the cart',
+              errors: [{ message: requiredErrorMessage }],
+              perform,
+           }
+         : {
+              behavior: 'allow',
+              perform,
```
Comment (danielcliu-ifixit) [inline extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx:63]: Do we need this perform?

**Comment by danielcliu-ifixit** [comment]
CR :v:  dev_block :v: on reviewing a few comments

File: extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx (hunk 3 above)
Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx:63]: Yes, it removes the error message and allows the buyer to progress if they don't want to agree and instead choose to remove the samsung product from their cart.

File: extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx (hunk 2 above)
Comment (masonmcelvain) [reply extensions/ui-extension/src/targets/purchase/checkout/delivery-address/SamsungConsent.tsx:67]: React doesn't allow hooks to be called conditionally
https://react.dev/warnings/invalid-hook-call-warning#breaking-rules-of-hooks

**Comment by masonmcelvain** [comment]
un_dev_block 👍🏻

**Review by danielcliu-ifixit — APPROVED** [review]
CR :v:

**Comment by deltuh-vee** [comment]
Carry over QA 🎬

**Pull request merged**